ass/document.py

```diff
@@ -578,3 +432,247 @@ class Command(_Event):
     """ A command event. Not widely supported.
     """
     TYPE = "Command"
+
+class CaseInsensitiveOrderedDict(collections.abc.MutableMapping):
+    """A case insensitive ordered dictionary that preserves the original casing."""
+
+    def __init__(self, *args, **kwargs):
+        self._dict = collections.OrderedDict(*args, **kwargs)
+        self._case_mapping = {key.lower(): key for key in self._dict}
+
+        if len(self._case_mapping) != len(self._dict):
+            raise ValueError("Duplicate keys provided for case insensitive dict")
+
+    def __getitem__(self, key):
+        return self._dict[self._case_mapping[key.lower()]]
+
+    def __setitem__(self, key, value):
+        if key.lower() not in self._case_mapping:
+            self._case_mapping[key.lower()] = key
+        self._dict[self._case_mapping[key.lower()]] = value
+
+    def __delitem__(self, key):
+        del self._dict[self._case_mapping[key.lower()]]
+        del self._case_mapping[key.lower()]
+
+    def __iter__(self):
+        return iter(self._dict)
+
+    def __len__(self):
+        return len(self._dict)
+
+    def __repr__(self):
+        return repr(self._dict)
+
+    def __str__(self):
+        return str(self._dict)
+
+class LineSection(collections.abc.MutableSequence):
+    FORMAT_TYPE = "Format"
+    LINE_PARSERS = None
+    FIELD_ORDER = None
```

Comment:
As I mentioned for `TYPE`, these shouldn't be allcaps if they are changed during runtime. Since this is new api, you might as well do it when introducing.

Reply:
I did this to be somewhat consistent with `TYPE`, but you're right that it's not very pretty. I'll change it.